### PR TITLE
Dlg Enter 

### PR DESCRIPTION
**Pull request by stevenndungu**
Dlg Enter (Not Date).
Added a preview pane on the design. Looking how I will implement that if the design is OK.

### CODE REVIEW

**Comment by dannyparsons** [comment]
Why have you called the pull request Open refine?

**Comment by dannyparsons** [comment]
You could have a look at how the preview is implemented in the Calculator, maybe looking at it with @FrancoisJRenaud to understand how that works, if he has time.
If not then I'm happy with not having the preview implemented yet.
Is there anything else in the dialog that still needs completing?

**Comment by stevenndungu** [comment]
Did it on the same branch I did OpenRefine. What I did forgot is change the name while making pull request on the git (that was a mistake).

Ok will check with Francois.

**Comment by dannyparsons** [comment]
Ok, you should definitely not be on the same branch. You should have deleted that branch when it was merged, and started on a new one. You can carry on for now, but remember for next time.
You can still rename this pull request now so it's more sensible.

**Comment by stevenndungu** [comment]
Ok, got you right thanks.

**Comment by dannyparsons** [comment]
A couple of points from testing quickly:
1. I think there should be a backspace button on the keyboard, I couldn't see one
2. When I reopened the dialog, the receiver clears and doesn't show the command I typed in before closing
3. The first time I clicked "LTRS" when the receiver was blank, it added "letters" instead of "LETTERS"
4. There are inconsistencies when the length of the output is not equal to the length of the data frame. If it's longer, it gives an error, if it's shorter but a divisor of the length of the data frame then it works, otherwise it gives an error. This is not to be fixed by this dialog (the same issue should appear in the calculator) as it is a property of the add columns method. But I'm highlighting it here so we think about how this should work in this dialog and others.

**Comment by FrancoisJRenaud** [comment]
1. Is the backspace not the del button ? Bottom left ?

**Comment by dannyparsons** [comment]
5. The receiver clears when I switch data frames. This is correct from the calculator, as it may contain column names, but I'm not sure we want this for this dialog, where the receiver content could be seen as less dependant on the data frame. @rdstern? Happy to have to resetting for now as it's safe.

**Comment by dannyparsons** [comment]
Yes, @FrancoisJRenaud thanks, don't know how I missed that.

**Comment by dannyparsons** [comment]
6.. None of the buttons are lined up.

File: instat/dlgEnter.vb [hunk 1]
```diff
@@ -264,4 +271,64 @@ Public Class dlgEnter
...
+        Dim strTempScript As String = ""
+        Dim strVecOutput As CharacterVector
+        Dim bIsAssigned As Boolean
+        Dim bToBeAssigned As Boolean
+        Dim strAssignTo As String
+        Dim strAssignToColumn As String
+        Dim strAssignToDataFrame As String
+
+        'First store the RSyntax settings temporarily, as these will be modified in the Try process. 
+        'Quetion: could we not use a clone RSyntax method ?
```
Comment (dannyparsons) [inline instat/dlgEnter.vb]: This should be improved yes, but is linked to bigger changes needed to RSyntax.

File: instat/dlgEnter.vb [hunk 2]
```diff
@@ -264,4 +271,64 @@ Public Class dlgEnter
...
+                ucrInputTryMessage.SetName("")
+            Else
+                strAttach = clsAttach.ToScript(strTempScript)
+                frmMain.clsRLink.RunInternalScript(strTempScript & strAttach, bSilent:=True)
+                ucrBase.clsRsyntax.RemoveAssignTo()
+                strOutPut = ucrBase.clsRsyntax.GetScript
+                strVecOutput = frmMain.clsRLink.RunInternalScriptGetOutput(strOutPut, bSilent:=True)
+                If strVecOutput IsNot Nothing Then
+                    If strVecOutput.Length > 1 Then
+                        'This is not working as expected. First of all Mid takes as arguments string, start end length. Here we give first entry of strVecOutput (which is a characterVector and not a string apparently..)? Then five would be the start ? Actually displays 19 letters... Does Mid has an overloaded definition I couldn't find ?
```
Comment (dannyparsons) [inline instat/dlgEnter.vb]: This is working how I expected when I tested. What's not working?
`Mid` does only take Strings and `strVecOutput` is a `CharacterVector` (a custom type from `RDotNet`), but each element of `strVecOutput` i.e. `strVecOutput(0)` is a String.
It doesn't show all the output, just the output from the first line you would get in the R console. So `strVecOutput(1)` would give you the next line. In most cases the first line is enough of a preview I think, and also avoids it being slow for long data frames.

**Comment by dannyparsons** [comment]
For 6.. there are alignment tools in Visual Studio, normally on one of the toolbars which you can use to align things automatically in different ways.

**Comment by stevenndungu** [comment]
Sorry, was off to some engagement in Eldoret and Kitale over the weekend
(seen your comments now). For;

   1.   As Francois mentioned.
   2. Something I should be checking tomorrow.
   3.  I corrected on the version I send on Friday.
   4.  Was discussing with Maxwell and Francois on Friday for some time but
   were yet to    agree on the method to use ultimately.

this is what I had in mind when dealing with R, I was enquiring if its
implementable.

[image: Inline image 1]

 What came up, was if we really wanted to to have NA's added whenever the
length is longer than the vector processed, in all the cases. For Instance,
if we wanted to add a new column with months on survey, it wont be helpful
- for months loop from Jan-Dec. But for seq,letters...could be helpful. Am
happy to get your suggestions as well.

   5 & 6. Will check on them and correct them.
 Thanks

**Comment by dannyparsons** [comment]
@stevenndungu What's the state of this? Are you still working on my suggestions?

**Comment by stevenndungu** [comment]
I have worked on all the other suggestions but  4) am yet to crack through..

**Comment by stevenndungu** [comment]
I wanted to make  a pull request of the whole suggestions made at ago.

**Comment by dannyparsons** [comment]
If you check what I said for 4:

> This is not to be fixed by this dialog (the same issue should appear in the calculator) as it is a property of the add columns method.

So you shouldn't have been spending any time trying to do this in this dialog. It needs to be done in the R code for general use.

**Comment by dannyparsons** [comment]
It's fine to send all your work at once, but for safety I strongly suggest you sync your branches online as often as possible. If your laptop breaks/gets stolen at least if you sync your branches online you don't lose any work you've done. I try to do this at least twice a day. This is a good reminder for everyone @africanmathsinitiative/instat

**Comment by dannyparsons** [comment]
Why have you deleted this? There's a lot of work here already.

**Comment by stevenndungu** [comment]
Ooh! sorry, mistaken it to import open refine.

**Comment by dannyparsons** [comment]
You can rename the pul request now to avoid confusion.

**Comment by stevenndungu** [comment]
Sure

**Pull request merged**